File: 3121-count-the-number-of-special-characters-ii/3121-count-the-number-of-special-characters-ii.go
func numberOfSpecialChars(word string) int {
    lowerLast := make([]int, 26)
    upperFirst := make([]int, 26)

    for i := 0; i < 26; i++ {
        lowerLast[i] = math.MaxInt
        upperFirst[i] = -1
    }

    for i, ch := range word {
        if ch >= 'a' && ch <= 'z' {
            lowerLast[ch-'a'] = i
        } else {
            ord := ch-'A'
            if upperFirst[ord] == -1 {
                upperFirst[ord] = i
            }
        }
    }

    ans := 0
    for i := 0; i < 26; i++ {
        if lowerLast[i] < upperFirst[i] {
            ans++
        }
    }
    return ans
}

File: 3121-count-the-number-of-special-characters-ii/3121-count-the-number-of-special-characters-ii.rs
impl Solution {
    pub fn number_of_special_chars(word: String) -> i32 {
        let mut lower_last: [i32; 26] = [i32::MAX; 26];
        let mut upper_first: [i32; 26] = [i32::MIN; 26];

        for (i, ch) in word.char_indices() {
            if ch >= 'a' && ch <= 'z' {
                let ord = ch as usize - 'a' as usize;
                lower_last[ord] = i as i32;
            } else {
                let ord = ch as usize - 'A' as usize;
                if upper_first[ord] == i32::MIN {
                    upper_first[ord] = i as i32;
                }
            }
        }

        let mut ans = 0;
        for i in 0..26 {
            if lower_last[i] < upper_first[i] {
                ans += 1;
            }
        }

        ans
    }
}